### DISCUSSION TITLE
How to work with extension-less file names?

### DISCUSSION THREAD

**Discussion by ivanjaros** [libvips/libvips#4261] [General]
I store files without their file extension. I do know their name and mime though. How can i tell vips that i want to
 `vips copy 123[filename=foo.jpg] 456[filename=foo.avif]` or `vips copy 123[mime=image/jpeg] 456[mime=image/avif]`?

**Comment by jcupitt** [comment]
It should just work -- libvips uses sniffers rather than extensions to select image loaders.

For example, you can write:

```
vips copy k2.jpg .png | vips copy stdin y.jpg
```

That makes the first copy write to stdout in PNG format, and the second read from stdin and write to JPG. The second vips is able to identify the filetype without a filename.

**Reply by ivanjaros** [reply]
calling `vips heifsave src test[compression=av1,encoder=aom]` where src is jpg image and test is supposed to be avif image gives me `heifsave: Unsupported compression`. same error that i get without adding any properties to the output file. I know what output i want so i can hard-code it but i do not want to be writing input switch for all image types to match extension-less files. anyway it seems output is problematic, not input.

**Reply by jcupitt** [reply]
You need:

```
vips heifsave src test --compression=av1 --encoder=aom
```

**Reply by jcupitt** [reply]
The libvips CLI usually executes a pipeline of three operations: a loader (selected from the magic number in the input data), the operation named first in the cli arg, and a saver selected from the output filename extension.

Any `--flag xxx` options are sent to the named operation, but that leaves no way to pass options to the loader and saver, so libvips uses the strange `[x=y]` syntax on filenames for this.

```
vips gravity x.jpg[shrink=2] y.tif[compression=lzw] centre 1000 1000 --background "255 0 0"
```

Executes something like:

```
jpegload x.jpg --shrink 2 | 
    gravity centre 1000 1000 --background "255 0 0" | 
        tiffsave y.tif --compression lzw
```

Having the load and save options in the filenames is often very convenient. You can write a little script to do something and not have to worry about file formats at all, just use `$1` or whatever. The person using your script can process GIF or AVIF or whatever they like just by adding filename options.

**Reply by ivanjaros** [reply]
this `vips heifsave src test --compression=av1 --encoder=aom` works only for this use case, if i want to perform any kind of action, like flip(`vipsflip src flip_y VIPS_DIRECTION_VERTICAL --compression=av1 --encoder=aom`) it errors with unknown arguments.

**Reply by ivanjaros** [reply]
right now i have to use file extension and then rename the file afterwards.

**Reply by jcupitt** [reply]
If you write to a suffix with no filename, it'll write to stdout. So you could do:

```
vips flip xxxx .avif[compression=av1,encoder=aom] horizontal > yyyy
```

**Reply by ivanjaros** [reply]
i might use that in the future, but for now using the renaming seems the way to go.
